pkg/controller/core/workload_controller.go

```diff
@@ -170,6 +171,16 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 
 	finishedCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadFinished)
 	if finishedCond != nil && finishedCond.Status == metav1.ConditionTrue {
+		// Unset quota reservation for finished workloads
+		if workload.HasQuotaReservation(&wl) {
+			log.V(2).Info("Unsetting quota reservation for finished workload")
+			if err := workload.PatchAdmissionStatus(ctx, r.client, &wl, r.clock, func() (*kueue.Workload, bool, error) {
+				return &wl, workload.UnsetQuotaReservationWithCondition(&wl, kueue.WorkloadFinished, "Workload finished", r.clock.Now()), nil
+			}); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to unset quota reservation for finished workload: %w", err)
+			}
+		}
```

Comment:
Instead of using an extra request to do it from workload_controller it should be done as part of the request which sets Finished condition. So, I imagine we have one helper function workload.Finish, called from workloadslicing.Finish, and reconciler.go (potentially other places), This function will set the Finished=True condition, and also set QuotaReserved=False.